src/Futhark/CLI/Eval.hs

```diff
@@ -0,0 +1,150 @@
+module Futhark.CLI.Eval (main) where
+
+import Control.Exception
+import Control.Monad
+import Control.Monad.Except
+import Control.Monad.Free.Church
+import Data.Maybe
+import Data.Text qualified as T
+import Data.Text.IO qualified as T
+import Futhark.Compiler
+import Futhark.MonadFreshNames
+import Futhark.Pipeline
+import Futhark.Util (toPOSIX)
+import Futhark.Util.Options
+import Futhark.Util.Pretty
+import Language.Futhark
+import Language.Futhark.Interpreter qualified as I
+import Language.Futhark.Parser
+import Language.Futhark.Semantic qualified as T
+import Language.Futhark.TypeChecker qualified as I
+import Language.Futhark.TypeChecker qualified as T
+import System.Exit
+import System.FilePath
+import System.IO
+import Prelude
+
+main :: String -> [String] -> IO ()
+main = mainWithOptions interpreterConfig options "options... <exprs...>" run
+  where
+    run [] _ = Nothing
+    run exprs config = Just $ runExprs exprs config
+
+runExprs :: [String] -> InterpreterConfig -> IO ()
+runExprs exprs cfg = do
+  let InterpreterConfig _ file = cfg
+  maybe_new_state <- newFutharkiState cfg file
+  (src, env, ctx) <- case maybe_new_state of
+    Left _ -> do
+      hPutStrLn stderr $ fromJust file <> ": file not found."
+      exitWith $ ExitFailure 2
+    Right s -> pure s
+  runExprs' exprs src env ctx
+
+runExprs' :: [String] -> VNameSource -> T.Env -> I.Ctx -> IO ()
+runExprs' [] _ _ _ = pure ()
+runExprs' (x : xs) src env ctx = do
+  runExpr src env ctx x
+  runExprs' xs src env ctx
+
+-- Use parseExp, checkExp, then interpretExp.
+runExpr :: VNameSource -> T.Env -> I.Ctx -> String -> IO ()
+runExpr src env ctx str = do
+  uexp <- case parseExp "" (T.pack str) of
+    Left (SyntaxError _ serr) -> do
+      T.hPutStrLn stderr serr
+      exitWith $ ExitFailure 1
+    Right e -> pure e
+  fexp <- case T.checkExp [] src env uexp of
+    (_, Left terr) -> do
+      hPutDoc stderr $ I.prettyTypeError terr
+      exitWith $ ExitFailure 1
+    (_, Right (_, e)) -> pure e
+  let ext = I.interpretExp ctx fexp
+  pval <- runInterpreter' ext
```

Comment:
No need to bind `ext`, just do `pval <- runInterpreter' $ I.interpretExp ctx fexp`.